### ISSUE TITLE
The length and width of prediction frames for objects of the same size are too different

### ISSUE THREAD

**Issue by venxzw** [ultralytics/ultralytics#20081]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

The picture scaling size of my training set was 640×640, and all objects of different sizes were marked with square boxes during data preparation. After successful training of 300epochs, the trained model was used to predict pictures with the same size of objects. Unfortunately, the length and width of the prediction boxes for some objects were too different.Since I need to utilize the center point of the prediction box for subsequent processing, I want to know what is causing this problem. In addition, I have missed detection after converting the trained pt model to the onnx model, but using the pt model does not cause this problem.

### Environment

Environment windows 11
python=3.8.20
torch=2.01
cuda=11.8
ultralytics was downloaded in October 2024 for version 8.3.19

### Minimal Reproducible Example

I use the defalut.yaml command line to inference like yolo task=detect mode=predict model=default.yaml

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @venxzw, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to report this issue. An Ultralytics engineer will assist you shortly. In the meantime, here are some suggestions and resources to help troubleshoot:

### Suggestions

If this is a 🐛 Bug Report:
- Please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/). Based on your description, it seems the provided command might not fully reproduce the issue. Try including a small, sharable dataset or a detailed script that replicates the prediction inconsistency and ONNX conversion issue.

If this is related to custom training:
- Verify that your dataset annotations are consistent with our [dataset formatting guidelines](https://docs.ultralytics.com/datasets/). Irregular annotation formats can lead to unexpected behavior.
- Ensure you’ve followed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for optimal performance.

For the ONNX conversion issue, consider testing the exported model using our [ONNX export and inference workflows](https://docs.ultralytics.com/tasks/export/) to confirm compatibility.

### Upgrade

It seems you are using an older version (`8.3.19`). To ensure you benefit from the latest fixes and features, please upgrade to the latest release of the `ultralytics` package along with its dependencies. Use the following command in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

### Environments

You can also test YOLO in verified environments that come pre-configured with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Resources

- For extensive documentation, visit our [Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples.
- For real-time discussions, join our [Discord](https://discord.com/invite/ultralytics) 🎧.
- Participate in in-depth conversations on [Discourse](https://community.ultralytics.com/) or check out our [Subreddit](https://reddit.com/r/Ultralytics).

### Status

To check if this issue might already be resolved in the latest version, verify our CI status below:
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

We’re here to help! Please provide any additional information that might assist in debugging, and we’ll address your concerns as quickly as possible. Thank you again for contributing to Ultralytics 🌟!

**Comment by glenn-jocher** [comment]
Hi @venxzw,

This discrepancy is expected behavior. During training, YOLO learns to predict the best-fitting bounding boxes around objects, even if your training annotations were square. The model optimizes for accuracy rather than preserving the original box shape.

If you need square predictions for your application, you can post-process the bounding boxes. Here's how to get the center point and convert to a square box:

```python
# Get center point from predictions
results = model(img)
for r in results:
    boxes = r.boxes.xywh  # Get boxes in xywh format
    for box in boxes:
        x_center, y_center, width, height = box[:4]
        # Create square box with max dimension
        square_size = max(width, height)
        # New square box coordinates (xywh format)
        square_box = [x_center, y_center, square_size, square_size]
```

Regarding the ONNX conversion issue with missed detections, this is a separate problem that may require different confidence thresholds or export settings. Could you share your export command and more details about the missed detections?

**Comment by glenn-jocher** [comment]
Thank you for the update @venxzw. I should clarify that adding anchor boxes isn't an option for YOLO11 as it's designed as an anchor-free detector (unlike YOLOv5). This architectural difference brings advantages in detection accuracy and flexibility, but it means the model will naturally predict boxes that best fit object shapes regardless of annotation style.

The post-processing approach I suggested remains the most straightforward solution for enforcing square boxes. It's good to hear you've identified the cause of your ONNX conversion issue - parameter settings during negative sample fine-tuning can indeed affect detection behavior across different export formats.

**Comment by venxzw** [comment]
> Thank you for the update [@venxzw](https://github.com/venxzw). I should clarify that adding anchor boxes isn't an option for YOLO11 as it's designed as an anchor-free detector (unlike YOLOv5). This architectural difference brings advantages in detection accuracy and flexibility, but it means the model will naturally predict boxes that best fit object shapes regardless of annotation style.
> 
> The post-processing approach I suggested remains the most straightforward solution for enforcing square boxes. It's good to hear you've identified the cause of your ONNX conversion issue - parameter settings during negative sample fine-tuning can indeed affect detection behavior across different export formats.
@glenn-jocher  I understand your point, but what I want to convey is that post-processing cannot solve the problem of the target center point being offset. Even if I adjust the prediction box to a square, the center point has already shifted. In this case, my prediction box is larger than the target object, and the obtained center point is biased to the left of the actual center point of the target object. I'm not sure if it's a problem with the bounding box being too large when preparing the dataset. Additionally, I know that YOLOv11 is anchor-free. I'm not quite sure if it can be changed to anchor-based to solve this problem.

**Comment by glenn-jocher** [comment]
@venxzw I understand your concern about the center point offset issue. This is actually more related to your dataset annotations than the anchor-free architecture. If your training bounding boxes were consistently larger than objects and/or biased to one side, the model would learn this pattern during training.

To address this issue:
1. Review your annotation process to ensure boxes are properly centered on objects
2. Consider using normalized box coordinates (`r.boxes.xywhn`) to verify center point consistency
3. You may need to re-annotate with more precise bounding boxes if there was systematic bias

YOLO11 can't be converted to anchor-based as it's a fundamental architectural difference. Even with anchor-based models, imprecise annotations would still result in biased predictions. The solution lies in improving annotation quality rather than changing the model architecture.

**Comment by venxzw** [comment]
> [@venxzw](https://github.com/venxzw) I understand your concern about the center point offset issue. This is actually more related to your dataset annotations than the anchor-free architecture. If your training bounding boxes were consistently larger than objects and/or biased to one side, the model would learn this pattern during training.
> 
> To address this issue:
> 
> 1. Review your annotation process to ensure boxes are properly centered on objects
> 2. Consider using normalized box coordinates (`r.boxes.xywhn`) to verify center point consistency
> 3. You may need to re-annotate with more precise bounding boxes if there was systematic bias
> 
> YOLO11 can't be converted to anchor-based as it's a fundamental architectural difference. Even with anchor-based models, imprecise annotations would still result in biased predictions. The solution lies in improving annotation quality rather than changing the model architecture.

@glenn-jocher Thank you very much for your reply. I have solved the problem by re-labeling the target, specifically by making the bounding box exactly the size of the target.

**Comment by glenn-jocher** [comment]
Great to hear you've resolved the issue @venxzw! Your solution of re-labeling to make the bounding boxes match the exact size of the targets is exactly right. This confirms what I suspected - annotation quality is crucial for getting precise center point predictions, especially with anchor-free detectors like YOLO11.

This experience highlights the importance of high-quality annotations in training data, as mentioned in our [troubleshooting guide](https://docs.ultralytics.com/guides/yolo-common-issues/). Anchor-free detection offers advantages in flexibility and accuracy, but it does require consistent and precise labeling.

Thanks for sharing your solution - it will be helpful for others facing similar challenges.
